tool/tctl/common/device/resource.go

```diff
@@ -0,0 +1,154 @@
+	Credential    *devicepb.DeviceCredential      `json:"credential,omitempty"`
+	CollectedData []*devicepb.DeviceCollectedData `json:"collected_data,omitempty"`
+}
+
+// checkAndSetDefaults sanity checks Resource fields to catch simple errors, and
+// sets default values for all fields with defaults.
+func (r *Resource) checkAndSetDefaults() error {
+	if err := r.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if r.Kind == "" {
+		r.Kind = types.KindDevice
+		// Sanity check.
+	} else if r.Kind != types.KindDevice {
+		return trace.BadParameter("unexpected resource kind %q, must be %q", r.Kind, types.KindDevice)
+	}
+
+	if _, err := devicetrust.ResourceOSTypeFromString(r.Spec.OsType); err != nil {
+		trace.Wrap(err)
+	}
+
+	if r.Spec.AssetTag == "" {
+		return trace.BadParameter("missing asset tag")
+	}
+
+	if _, ok := devicepb.DeviceEnrollStatus_value[r.Spec.EnrollStatus]; !ok {
+		r.Spec.EnrollStatus = devicepb.DeviceEnrollStatus_name[int32(devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_UNSPECIFIED)]
+	}
+
+	return nil
+}
+
+// UnmarshalDevice parses a device in the Resource format which matches
+// the expected YAML format for Teleport resources, sets default values, and
+// converts to *devicepb.Device.
+func UnmarshalDevice(raw []byte) (*devicepb.Device, error) {
+	var resource Resource
+	if err := utils.FastUnmarshal(raw, &resource); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := resource.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return resourceToProto(&resource)
+}
+
+// ProtoToResource converts a *devicepb.Device into a *Resource which
+// implements types.Resource and can be marshaled to YAML or JSON in a
+// human-friendly format.
+func ProtoToResource(device *devicepb.Device) *Resource {
+	r := &Resource{
+		ResourceHeader: types.ResourceHeader{
+			Kind:    types.KindDevice,
+			Version: device.ApiVersion,
+			Metadata: types.Metadata{
+				Name: device.Id,
+			},
+		},
+		Spec: ResourceSpec{
+			OsType:        devicetrust.ResourceOSTypeToString(device.OsType),
+			AssetTag:      device.AssetTag,
+			EnrollToken:   device.EnrollToken,
+			EnrollStatus:  devicepb.DeviceEnrollStatus_name[int32(device.EnrollStatus)],
+			Credential:    device.Credential,
+			CollectedData: device.CollectedData,
+		},
+	}
+
+	if device.CreateTime != nil {
+		r.Spec.CreateTime = device.CreateTime.AsTime()
+	}
+
+	if device.UpdateTime != nil {
+		r.Spec.UpdateTime = device.UpdateTime.AsTime()
+	}
+
+	return r
+}
+
+func resourceToProto(r *Resource) (*devicepb.Device, error) {
+	osType, err := devicetrust.ResourceOSTypeFromString(r.Spec.OsType)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	dev := &devicepb.Device{
+		ApiVersion:    r.Version,
+		Id:            r.Metadata.Name,
+		OsType:        osType,
+		AssetTag:      r.Spec.AssetTag,
+		EnrollToken:   r.Spec.EnrollToken,
+		EnrollStatus:  devicepb.DeviceEnrollStatus(devicepb.DeviceEnrollStatus_value[r.Spec.EnrollStatus]),
```

Comment:
We can omit this altogether, since we are likely to mostly throw away the values anyway.

I'd leave a comment saying it's intentional, though.

Reply:
What exactly are you referring to here? Throw out which values? Where?

Reply:
I'm referring to the EnrollStatus field.

Reply:
Resolving as per slack conversation.